canals/component/input_output.py

```diff
@@ -2,78 +2,151 @@
 #
 # SPDX-License-Identifier: Apache-2.0
 import logging
-import inspect
-from dataclasses import fields
+from enum import Enum
+from dataclasses import fields, is_dataclass, dataclass, asdict, MISSING
+
+from canals.errors import ComponentError
 
 logger = logging.getLogger(__name__)
 
 
-class BaseIODataclass:  # pylint: disable=too-few-public-methods
+def _make_fields_optional(class_: type):
     """
-    Base class for input and output classes of components.
+    Takes a dataclass definition and modifies its __init__ so that all fields have
+    a default value set.
+    If a field has a default factory use it to set the default value.
+    If a field has neither a default factory or value default to None.
     """
+    defaults = []
+    for field in fields(class_):
+        default = field.default
+        if field.default is MISSING and field.default_factory is MISSING:
+            default = None
+        elif field.default is MISSING and field.default_factory is not MISSING:
+            default = field.default_factory()
+        defaults.append(default)
+    # mypy complains we're accessing __init__ on an instance but it's not in reality.
+    # class_ is a class definition and not an instance of it, so we're good.
+    # Also only I/O dataclasses are meant to be passed to this function making it a bit safer.
+    class_.__init__.__defaults__ = tuple(defaults)  # type: ignore
```

Comment:
A class definition is not an `object` but a `type` so that type hint is technically correct. There is currently no way to set the type hint to be of a class definition.

```
In [68]: class MyClass:
    ...:     pass
    ...:

In [69]: type(MyClass)
Out[69]: type
```

Changing to `object` still doesn't make `mypy` happy.

```
canals/component/input_output.py:21: error: Argument 1 to "fields" has incompatible type "object"; expected "Union[DataclassInstance, Type[DataclassInstance]]"  [arg-type]
canals/component/input_output.py:31: error: Accessing "__init__" on an instance is unsound, since instance.__init__ could be from an incompatible subclass  [misc]
```

Reply:
I found this: https://mypy.readthedocs.io/en/stable/kinds_of_types.html#the-type-of-class-objects Not 100% it's applicable given that there's no base class, but worth knowing for later maybe

Reply:
It might as well be that `Type` works while `type` doesn't